garden_ai/mlmodel.py

```diff
@@ -0,0 +1,36 @@
+import pickle
+import mlflow  # type: ignore
+from typing import List
+
+
+class ModelUploadException(Exception):
+    """Exception raised when an attempt to upload a model to ML Flow fails"""
+    pass
+
+
+def upload_model(model_path: str, model_name: str, user_email: str, extra_pip_requirements: List[str] = None):
```

Comment:
on the other hand, I don't know if downloading a model should/could be attached to a garden client, since pipeline code wouldn't have a garden client instance when they want to download their model, and it might be odd for upload_model to be attached to a client if download_model was not

Reply:
This is low-key a deep question and something I'd like to think about iteratively. Maybe attaching it to the GardenClient will be the right move.

I have an inkling that we should be trying to separate things from the GardenClient. Or at least find a clean way to separate the logic that manages logging in from logic that _does stuff_ like scaffolding pipeline code or what have you. I wanted to write a test for this function that didn't have to mock out a GardenClient because I don't really want to think about setting up Globus login when testing model upload.

There's a lot of ways we can create that kind of "seam" though. This is the flavor of thing I want to brainstorm with you next week. For now I'll leave it this way if that's chill.

Reply:
yeah totally agree I think it's worth a brainstorm